edgelet/edgelet-http/src/certificate_manager.rs

```diff
@@ -0,0 +1,243 @@
+
+        let mut ca_certs = Stack::new().with_context(|_| ErrorKind::CertificateConverstionError)?;
+        for cert in certs.split_off(1) {
+            ca_certs
+                .push(cert)
+                .with_context(|_| ErrorKind::CertificateConverstionError)?;
+        }
+
+        let key = PKey::private_key_from_pem(stored_cert_bundle.1.as_bytes())
+            .expect("Issue getting private key");
+
+        let server_cert = &certs[0];
+        let mut builder = Pkcs12::builder();
+        builder.ca(ca_certs);
+        let pkcs_certs = builder
+            .build("", "", &key, &server_cert)
+            .with_context(|_| ErrorKind::CertificateConverstionError)?;
+
+        Ok(pkcs_certs
+            .to_der()
+            .with_context(|_| ErrorKind::CertificateConverstionError)?)
+    }
+
+    pub fn get_certificate(&self) -> Result<(String, String), Error> {
+        // First, try to directly read
+        {
+            let cert = self
+                .certificate
+                .read()
+                .expect("Locking the certificate for read failed.");
+
+            if let Some(cert) = cert.as_ref() {
+                return Ok(cert.clone());
+            }
+        }
+
+        // No valid cert so must create
+        let mut cert = self
+            .certificate
+            .write()
+            .expect("Locking the certificate for write failed.");
+
+        if let Some(cert) = cert.as_ref() {
+            Ok(cert.clone())
+        } else {
+            let (new_cert, private_key) = self
+                .create_cert()
+                .with_context(|_| ErrorKind::CertificateCreationError)?;
+            Ok(cert.get_or_insert((new_cert, private_key)).clone())
+        }
+    }
+
+    fn create_cert(&self) -> Result<(String, String), Error> {
+        let cert = self
+            .crypto
+            .create_certificate(&self.props)
+            .with_context(|_| ErrorKind::CertificateCreationError)?;
+
+        let cert_pem = cert
+            .pem()
+            .with_context(|_| ErrorKind::CertificateCreationError)?;
+
+        let cert_private_key = cert
+            .get_private_key()
+            .with_context(|_| ErrorKind::CertificateCreationError)?;
+
+        let pk = match cert_private_key {
+            Some(pk) => pk,
+            None => panic!("did not expect reference private key"),
+        };
+
+        let pk_bytes = match pk {
+            PrivateKey::Ref(_) => panic!("did not expect reference private key"),
+            PrivateKey::Key(KeyBytes::Pem(k)) => k,
+        };
+
+        let cert_str = String::from_utf8(cert_pem.as_ref().to_vec())
+            .with_context(|_| ErrorKind::CertificateCreationError)?;
+
+        let key_str = String::from_utf8(pk_bytes.as_bytes().to_vec())
+            .with_context(|_| ErrorKind::CertificateCreationError)?;
+
+        Ok((cert_str, key_str))
+    }
+
+    #[cfg(test)]
+    fn has_certificate(&self) -> bool {
```

Comment:
Were you going to move this function into the test module below?

Reply:
Will follow up on this - I use this function to tell if my get_certificate code is working since if no certificate exists, it will create one. It operates on the Cert Manager struct and AFAIK, it must be where it is (unless I am missing some rust test code magic :) ) .

Reply:
I see. That's fine then.